crates/holochain_types/src/app.rs

```diff
@@ -27,7 +46,9 @@ pub struct InstallAppPayload {
 #[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
 pub struct InstallAppDnaPayload {
     /// The path of the DnaFile
-    pub path: PathBuf,
+    pub path: Option<PathBuf>, // This is a deprecated field should register dna contents via register
+    /// The path of the DnaFile
+    pub hash: Option<DnaHash>, // When path is removed, this will become non-opt
```

Comment:
I think you should use the `DnaSource` here instead of two options as `None, None` and `Some, Some` is invalid but cannot be expressed using the enum.
You could make dna source a generic with a default like:
```rust
pub enum DnaSource<D: DnaFile> {
    Path(PathBuf),
    Wasm(D),
}
```
To use it in both places

Reply:
I would tend to agree in general about disallowing invalid representations, but I think it's fine here. This is an easy way to achieve backwards compatibility, and the errors from `None, None` and `Some, Some` are basically in the same class as deserialization errors, which are unavoidable and happen when people use the API incorrectly.